### DISCUSSION TITLE
@UseRequestContext with .createQueryBuilder method. @Cron

### DISCUSSION THREAD

**Discussion by sergeyampo** [mikro-orm/mikro-orm#2808] [Q&A] [answered]
Is it possible to use @UseRequestContext with EntityManager.createQueryBuilder from @mikro-orm/postgresql?

**Comment by B4nan** [comment, marked as answer]
`@UseRequestContext` is there to create contexts, its up to you what you do inside the handler

**Reply by sergeyampo** [reply]
> We can use the @UseRequestContext() decorator. It requires you to first inject the MikroORM instance to current context, it will be then used to create the context for you. Under the hood, the decorator will register new request context for your method and execute it inside the context. 
https://mikro-orm.io/docs/usage-with-nestjs#request-scoped-handlers-in-queues
I suppose I have to inject MikroORM from @mikroorm/core but this way I can't create queryBuilder.
```
import { MikroORM, UseRequestContext } from '@mikro-orm/core';
@Injectable()
export class FilesCronService {
  constructor(
    private readonly orm: MikroORM,
  ) {}
  @UseRequestContext()
  @Cron(CronExpression.EVERY_MINUTE)
  async something(): Promise<void> {
    there's no way for me here to get queryBuilder I have to inject EntityManager from @mikroorm/postgresql 
  }
}
```

**Reply by B4nan** [reply]
Oh I see, this is just about TS, you can just cast, the right instance will be there with the method (you can see that if you console.log it). 

Or alternatively you can even use `this.orm.em` and use the type argument of `MikroORM` class:

```
import { MikroORM, UseRequestContext } from '@mikro-orm/core';
import { PostgreSqlDriver } from '@mikro-orm/postgresql';

@Injectable()
export class FilesCronService {
  constructor(
    private readonly orm: MikroORM<PostgreSqlDriver>, // <= here
  ) {}
  @UseRequestContext()
  @Cron(CronExpression.EVERY_MINUTE)
  async something(): Promise<void> {
    this.orm.em.createQueryBuilder(...) // <= works now
  }
}

**Comment by sergeyampo** [comment]
It's interesting that scheduling only works in a such order:
```
  @Cron(CronExpression.EVERY_5_MINUTES)
  @UseRequestContext()
async handler(){ ...
```
Not this way:
```
  @UseRequestContext()
  @Cron(CronExpression.EVERY_5_MINUTES)
async handler(){ ...
```

**Reply by sergeyampo** [reply]
@B4nan Does it need to describe such behavior in the docs?

**Reply by B4nan** [reply]
The other way dont make much sense, imagine there would be two separate methods, each one with a single decorator and calling the other one:

```ts
@Cron(CronExpression.EVERY_5_MINUTES)
async handler(){
  this.context();
}

@UseRequestContext()
async context(){
  // ...
}
```

This is the supported way, looks correct, right?

Now the other way that does not work:

```ts
@UseRequestContext()
async context(){
  this.handler();
}

@Cron(CronExpression.EVERY_5_MINUTES)
async handler(){
  // ...
}
```

But this now have a different meaning - you would create only one context, and reuse it in all the CRON jobs.

(at least that is how I understood it, haven't really debugged anything so I might be wrong)

**Reply by B4nan** [reply]
Note that this is actually already part of the docs:

> Another thing to look out for how you combine them with other decorators. For example if you use it in combination with NestJS's "[BullJS queues module](https://docs.nestjs.com/techniques/queues)", a safe bet is to extract the part of the code that needs a clean [docs](https://mikro-orm.io/docs/identity-map), either in a new method or inject a separate service.

https://mikro-orm.io/docs/usage-with-nestjs#request-scoped-handlers-in-queues

Could be cleaner for sure, feel free to PR this.
